pretty/examples_test.go

```diff
@@ -308,3 +308,80 @@ func ExampleCompare_debugging() {
 	// + Stolen: false,
 	//  }
 }
+
+type ListNode struct {
+	Value int
+	Next  *ListNode
+}
+
+func circular(nodes int) *ListNode {
+	final := &ListNode{
+		Value: nodes,
+	}
+	final.Next = final
+
+	recent := final
+	for i := nodes - 1; i > 0; i-- {
+		n := &ListNode{
+			Value: i,
+			Next:  recent,
+		}
+		final.Next = n
+		recent = n
+	}
+	return recent
+}
+
+func ExamplePrint_defaultRecursivePrinting() {
+	pretty.Print(circular(3))
+
+	// Output:
+	// <1>:
+	// {Value: 1,
+	//  Next:  {Value: 2,
+	//          Next:  {Value: 3,
+	//                  Next:  -> <1>}}}
+}
+
+func ExampleConfig_recursiveDiffable() {
+	config := &pretty.Config{
+		Diffable: true,
+	}
+	config.Print(circular(3))
+
+	// Output:
+	// <1>:
+	// {
+	//  Value: 1,
+	//  Next: {
+	//   Value: 2,
+	//   Next: {
+	//    Value: 3,
+	//    Next: -> <1>,
+	//   },
+	//  },
+	// }
+}
+
+func ExampleConfig_Compare_selfReferential() {
+	got, want := circular(3), circular(4)
+	fmt.Printf("Diff: (-got +want)\n%s", pretty.Compare(got, want))
+
+	// Output:
+	// Diff: (-got +want)
+	//  <1>:
+	//  {
+	//   Value: 1,
+	//   Next: {
+	//    Value: 2,
+	//    Next: {
+	//     Value: 3,
+	// -   Next: -> <1>,
+	// +   Next: {
+	// +    Value: 4,
+	// +    Next: -> <1>,
+	// +   },
+	//    },
+	//   },
+	//  }
+}
```

Comment:
I haven't looked at your point tracking logic, but what would happen in the case of a sibling pointing to another sibling in a graph structure (where there is no cyclic dependency, but where a node is visited multiple times)?

Your code might not even have a problem with this; this example just came to mind while looking at the examples here. But, if this is an issue, I'd suggest the second output in my example:
 
```
// Example ASCII graph:
//   1 ---------> 2 --> 4
//     \-->3--/

// Would it be:
{
  Value: 1,
  <2>
  Left: {
    Value: 2
    Left: {
      Value: 4,
      Left: nil,
      Right: nil,
  },
  Right: {
    Value: 3,
    Left: -> <2>
    Right: nil
  }
}



// or
{
  Value: 1,
  Left: {
    Value: 2
    Left: {
      Value: 4,
      Left: nil,
      Right: nil,
    },
    Right: nil,
  },
  Right: {
    Value: 3,
    Left: {
      Value: 2
      Left: {
        Value: 4,
        Left: nil,
        Right: nil,
      },
      Right: nil,
    },
    Right: nil
  }
}
```

Reply:
@crunk1 - it is the second version because @kylelemons wanted just duplicate non-cyclic references to be duplicates. Please see discussion on this here:

https://github.com/kylelemons/godebug/pull/17#issuecomment-323615023
https://github.com/kylelemons/godebug/pull/17#issuecomment-323546349

(on my side I would prefer we always, or via option, deduplicate even non-cyclic references).

@kylelemons it is pity this whole topic got stuck...

Reply:
In the mean time, I've found this to work pretty well (not meaning to advertise other repos):

```
import(
	"github.com/davecgh/go-spew/spew"
	"github.com/kylelemons/godebug/diff"
)

func doDiff(x, y interface{}) string {
    return diff.Diff(spew.Sdump(x), spew.Sdump(y))
}
```